src/machine/machined-core.c

```diff
@@ -456,3 +456,22 @@ int manager_acquire_image(Manager *m, const char *name, Image **ret) {
         TAKE_PTR(image);
         return 0;
 }
+
+int rename_image_and_update_cache(Manager *m, Image *image, const char* new_name) {
+        assert(m);
+        assert(image);
+        assert(new_name);
+
+        /* The image is cached with its name, hence it is necessary to remove from the cache before renaming. */
+        assert_se(hashmap_remove_value(m->image_cache, image->name, image));
+
+        int r = image_rename(image, new_name);
+        if (r < 0) {
+                image_unref(image);
+                return r;
+        }
+
+        /* Then save the object again in the cache. */
+        assert_se(hashmap_put(m->image_cache, image->name, image) > 0);
```

Comment:
this can fail because it involves a memory allocation. I understand this is a pre-existing bug, but if you touch this please fix this. given the cache is a "best effort" thing anyway, maybe on any kind of error just debug or warn log, free the cache, and release the image object and continue.

Reply:
Is this what you're thinking about?

```
          /* Then save the object again in the cache. */
          r = hashmap_put(m->image_cache, image->name, image);
          if (r < 0) {
                  image_unref(image);
                  return log_debug_errno(r, "Failed to put image into cache: %m");
          }
```

Reply:
and i think we should do `image = image_unref(image)`, just as a matter of hygiene. even if we never use `image` afterwards, it's a lot cleaner if we invalidate pointers that we dropped our ref of.